include/seqan3/io/detail/magic_header.hpp

```diff
@@ -13,83 +13,94 @@
 #pragma once
 
 #include <array>
+#include <tuple>
 
+#include <seqan3/core/type_traits/template_inspection.hpp>
 #include <seqan3/core/platform.hpp>
+#include <seqan3/core/type_list.hpp>
 
 namespace seqan3::detail
 {
 
-//!\brief Defines a magic byte sequence to disambiguate different compression formats. Default is empty.
-//!\ingroup io
-template <typename header_tag_t>
-inline constexpr std::array<char, 0> magic_header{};
-
 //!\brief A tag signifying a gz compressed file.
 //!\ingroup io
 struct gz_compression
-{};
+{
+    //!\brief The valid file extension for gz compression.
+    static inline const std::vector<std::string> file_extensions
+    {
+        {"gz"}
+    };
 
-/*!\brief The magic byte sequence to disambiguate gz compressed files.
- * \ingroup io
- *
- * \details
- *
- * Specialises seqan3::detail::magic_header for seqan3::detail::gz_compression.
- */
-template <>
-inline constexpr std::array<char, 3> magic_header<gz_compression>{'\x1f', '\x8b', '\x08'};
+    //!\brief The magic byte sequence to disambiguate gz compressed files.
+    inline static constexpr std::array<char, 3> magic_header{'\x1f', '\x8b', '\x08'};
+};
 
 //!\brief A tag signifying a bz2 compressed file.
 //!\ingroup io
 struct bz2_compression
-{};
+{
+    //!\brief The valid file extension for bz2 compression.
+    static inline const std::vector<std::string> file_extensions
+    {
+        {"bz2"}
+    };
 
-/*!\brief The magic byte sequence to disambiguate bz2 compressed files.
- * \ingroup io
- *
- * \details
- *
- * Specialises seqan3::detail::magic_header for seqan3::detail::bz2_compression.
- */
-template <>
-inline constexpr std::array<char, 3> magic_header<bz2_compression>{'\x42', '\x5a', '\x68'};
+    //!\brief The magic byte sequence to disambiguate bz2 compressed files.
+    inline static constexpr std::array<char, 3> magic_header{'\x42', '\x5a', '\x68'};
+};
 
 //!\brief A tag signifying a zstd compressed file.
 //!\ingroup io
 struct zstd_compression
-{};
+{
+    //!\brief The valid file extension for zstd compression.
+    static inline const std::vector<std::string> file_extensions
+    {
+        {"zst"}
+    };
 
-/*!\brief The magic byte sequence to disambiguate zstd compressed files.
- * \ingroup io
- *
- * \details
- *
- * Specialises seqan3::detail::magic_header for seqan3::detail::zstd_compression.
- */
-template <>
-inline constexpr std::array<char, 4> magic_header<zstd_compression>{'\x28', '\xb5', '\x2f', '\xfd'};
+    //!\brief The magic byte sequence to disambiguate zstd compressed files.
+    inline static constexpr std::array<char, 4> magic_header{'\x28', '\xb5', '\x2f', '\xfd'};
+};
 
 //!\brief A tag signifying a bgzf compressed file.
 //!\ingroup io
 struct bgzf_compression
-{};
+{
+    //!\brief The valid file extension for bgzf compression.
+    static inline const std::vector<std::string> file_extensions
+    {
+        {"bgzf"}
```

Comment:
For output we write bgzf format if we detect .gz extension. But the extension is not replaced. Thus there is no reason to put the gz extension here as well. The same goes for gz format which could also read bgzf files since they are compatible.